Add alignment string position support

alyst opened the pull request:
# Alignment string position support

## Types of changes

This PR implements the following changes:

* [x] :sparkles: New feature (A non-breaking change which adds functionality).
* [ ] :bug: Bug fix (A non-breaking change, which fixes an issue).
* [ ] :boom: Breaking change (fix or feature that would cause existing functionality to change).

## :clipboard: Additional detail

  - Mapping positions from/to the input or reference sequence onto the common alignment string is quite useful (e.g. when comparing annotations attached to specific positions of the aligned strings). In addition to `seq2ref()` and `ref2seq()`, this PR adds 4 new functions to the public API: `seq2agn`, `ref2agn`, `agn2seq`, `agn2ref`, which allow to convert between input/reference/alignment positions in any direction.

  - Technically, keeping track of alignment positions is implemented as an `AlignmentAnchor.agnpos` field. Updating the pairwise alignment algorithms to support alignment positions was easy, only the common traceback utility macros had to be changed. It also should not affect the performance or memory consumption.

  - I also did a few minor cleanups (`@inbounds` were beneficial, refactored position conversion to be more generic and using Julia 1.x approach (avoid generated functions, rely on constant propagation, argument splatting)).

  - **Provide a runnable example of use of your addition**. This lets reviewers
    and others try out the feature before it is merged or makes it's way to release.

## :ballot_box_with_check: Checklist

- [x] :art: The changes implemented is consistent with the [julia style guide](https://docs.julialang.org/en/stable/manual/style-guide/).
- [x] :blue_book: I have updated and added relevant docstrings, in a manner consistent with the [documentation styleguide](https://docs.julialang.org/en/stable/manual/documentation/).
- [ ] :blue_book: I have added or updated relevant user and developer manuals/documentation in `docs/src/`.
- [x] :ok: There are unit tests that cover the code changes I have made.
- [x] :ok: The unit tests cover my code changes AND they pass.
- [ ] :pencil: I have added an entry to the `[UNRELEASED]` section of the manually curated `CHANGELOG.md` file for this repository.
- [x] :ok: All changes should be compatible with the latest stable version of Julia.
- [x] :thought_balloon: I have commented liberally for any complex pieces of internal code.

alyst commented:
The failures due to `Registry.toml` problems, the testing even didn't start.
Also: is it possible to update CI to test on latest stable Julia and also on nightly (with *allow_failures*). I guess only LTS (is it 1.0 or 1.1?) Julia should be kept from the previous versions.

alyst commented:
Some other notes on the PR:
* In this PR I haven't yet updated the manual -- but if the maintainers think it's reasonable to include the alignment position functionality, I'll update it. The terminology (e.g. *alignment position*) also has to be clarified before committing to update the docs.
* The *PairwiseAlignment* iterator interface could be updated to return the alignment position. But that would break the API, so I don't know what to do about that.
* In addition to the alignment position, it would be nice to have API to query the input/reference sequence ranges aligned. Partly it exists as *Alignment.firstref/lastref*, but so far it's not public API and it doesn't cover input sequence positions.
* Not a part of this PR, but something that I noticed: there `Alignment`, `AlignedSequence` and `PairwiseAlignment` types. Judging from the names, `PariwiseAlignment` and `Alignment` are expected to be from the same type hierarchy, but in fact these are very different types. Perhaps, it's possible to rename `Alignment` into something like `AlignmentDefinition` or `AlignmentTransform` (and fix the .jl filename and field names of that type accordingly (e.g. into *transform* -- right now in some parts of the code one may read `aln.a.aln.anchors`)) to avoid the confusion.

alyst commented:
I don't know who would be the best to review, but since you were committing recently, maybe you can take a look, @BenJWard ? Or maybe you can suggest a more suitable reviewer? Thanks in advance!

alyst commented:
@BenJWard @jakobnissen If you would have time, could you please review this PR?

jakobnissen commented on src/operations.jl:
```diff
@@ -113,7 +113,7 @@ function Base.convert(::Type{Char}, op::Operation)
 
 function Base.print(io::IO, op::Operation)
-    write(io, convert(Char, op))
+    write(io, isvalid(op) ? convert(Char, op) : '?')
```
Should we support invalid operations? Perhaps it's a better design to not allow invalid operations to exist. I'm not sure though

alyst replied:
The time has passed, I don't know why I've added it. Maybe when I'll be revisiting the PR, I'll see why.
Otherwise I agree, it could be removed.

jakobnissen commented on src/alignment.jl:
```diff
@@ -100,41 +104,74 @@ function Base.show(io::IO, aln::Alignment)
-        refpos += i - anchor.seqpos
+    pos = destpos(anchor)
+    if ismatchop(anchor.op) ||
+        ((srcpos === agnpos) && ((destpos === seqpos) && isinsertop(anchor.op) || (destpos === refpos) && isdeleteop(anchor.op))) ||
+        ((destpos === agnpos) && ((srcpos === seqpos) && isinsertop(anchor.op) || (srcpos === refpos) && isdeleteop(anchor.op)))
+        pos += i - srcpos(anchor)
```
Might be better to use `==` here - if the type of any of the integers change, this code will needlessly break.

jakobnissen reviewed (commented):
Sorry for just getting around to it now. I can't wrap my head around the logic, but despite a few minor issues, it looks good.

I'm not 100% sure what this functionality is useful for, though. Does the concept of an "alignment position" make sense?

alyst commented on src/alignment.jl:
```diff
@@ -100,41 +104,74 @@ function Base.show(io::IO, aln::Alignment)
-Map a position `i` from sequence to reference.
-"""
-function seq2ref(aln::Alignment, i::Integer)::Tuple{Int,Operation}
-    idx = findanchor(aln, i, Val{true})
```
I was never sure whether `Val(...)` generates a code to allocate that object or not, so to avoid allocation I was using type approach, but I'll change it to `Val()`.

jakobnissen replied:
There is no allocation - the two things are completely equivalent. It's just the norm to use instances instead of types whenever you have a zero-field struct. (e.g. Julia's Base and the rest of BioJulia does it).

alyst replied:
Actually, that was not my code. %) In the refactored functions `Val` is replaced by passing function references.

alyst commented on src/alignment.jl:
```diff
@@ -100,41 +104,74 @@ function Base.show(io::IO, aln::Alignment)
+    idx = findanchor(aln, i, srcpos)
     if idx == 0
-        throw(ArgumentError("invalid sequence position: $i"))
+        if srcpos === seqpos
```
It's not a variable, it's a function defined in `anchors.jl` that retrieves a specific position.
That's why I'm using `===`. It relates to your other comments below.

The function references serve here both as functions that get the positions and as parameters that specify how counting should be done.
The motivation was to reduce code duplication that was there before (`seq2ref()` and `ref2seq()`),
because when introducing aligned position and mapping between all 3 types of positions, the code duplication would have been further increased.

alyst commented on src/alignment.jl:
```diff
@@ -172,65 +209,68 @@ function check_alignment_anchors(anchors)
-                return j
-            end
+        # invariant (activate this for debugging)
+        #@assert anchors[lo].$pos < i ≤ anchors[hi].$pos
```
I think it should be still relevant, because the function logic was not touched, I've just added support for aligned position + made it a normal function, not `@generated`.
I'd rather keep it here as a comment: it nicely documents the constraint with the algorithm, but the context might be performance-critical to enable it.
I'll update it to use the current way of getting the position (`pos(anchors...))`) and check internally (without pushing here) whether it doesn't throw in tests.

alyst replied:
So I checked and tests do not throw if the line is enabled.

alyst commented:
> I'm not 100% sure what this functionality is useful for, though. Does the concept of an "alignment position" make sense?

For multiple sequence alignments it definitely does, because it's actually the column of MSA.

But it also does make sense for pairwise one. If you want to display the pairwise alignment, you may need to know what is the position of a specific letter from either sequence in the aligned sequence.
My use case was to correctly display the post-translational modifications (phosphorylation, ubiquitination etc) of inidividual amino acids on the pairwise-aligned homologous protein sequences.
There could be regions that are present only on the first sequence or regions that are present only on the second one. So you definitely need an aligned position, which preserves everything.

alyst commented:
@jakobnissen  It looks like at the moment CI seems broken. Should this package be updated to use GitHab actions for CI?
I can try submitting the PR for that, but maybe it would be better if someone who has better knowledge of BioJulia ecosystem is doing that.

jakobnissen commented:
Yeah... BioJulia is strapped for dev manpower at the moment, so there is a lot of cleanup and housekeeping that needs to be done in all sorts of packages, and just haven't been done. If you make a PR, I'll merge it! You can look at e.g. BioSequences for how to do this.

jakobnissen commented:
@alyst Ping me when it's ready to be merged :)

alyst commented:
@jakobnissen Thank you! I think the PR is good to go.
I have addressed your suggestions, also I have added tests to the new *aln2xxx/xxx2aln()* methods that were not covered.
There are still new lines in the PR that are not covered, but they were derived from the existing exceptions that got more specific after *alnpos* was added.
They were not covered by the tests before. So I think it's fine to merge this PR, and address the coverage later by introducing more test cases.

jakobnissen commented:
Thanks for putting in the work!

timholy commented:
Docs need updating: https://github.com/BioJulia/BioAlignments.jl/blob/master/docs/src/alignments.md#representing-alignments
